### PR TITLE
Add OneOrMore and NOrMore types

### PR DESCRIPTION
**Pull request by darcyparker**
Depends on https://github.com/sindresorhus/type-fest/issues/284 being fixed.

This is still a draft. You can see I inlined @jonahsnider's improved `FixedLengthArray`. (It does not work with existing `FixedLengthArray`)

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -28,6 +28,7 @@ export {ConditionalPick} from './source/conditional-pick';
 export {UnionToIntersection} from './source/union-to-intersection';
 export {Stringified} from './source/stringified';
 export {FixedLengthArray} from './source/fixed-length-array';
+export {OneOrMore, NOrMore} from './source/n-or-more';
```
Comment (sindresorhus) [inline index.d.ts:31]: The naming on the second one is a bit unfortunate. I keep reading it as `no-r-more`.

File: readme.md [hunk 2]
```diff
@@ -121,6 +121,8 @@ Click the type names for complete docs.
 - [`FixedLengthArray`](source/fixed-length-array.d.ts) - Create a type that represents an array of the given type and length.
 - [`MultidimensionalArray`](source/multidimensional-array.d.ts) - Create a type that represents a multidimensional array of the given type and dimensions.
 - [`MultidimensionalReadonlyArray`](source/multidimensional-readonly-array.d.ts) - Create a type that represents a multidimensional readonly array of the given type and dimensions.
+- [`OneOrMore`](source/n-or-more.d.ts) - Create a type of one or more items.
```
Comment (sindresorhus) [inline readme.md:124]: I prefer using the word "elements" over "items", so we should use that consistency.

File: source/n-or-more.d.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/**
+Create a type of one or more items.
+
```
Comment (sindresorhus) [inline source/n-or-more.d.ts:3]: Need to mention use-cases and a usage example.

File: index.d.ts (hunk 1 above)
Comment (jonahsnider) [reply index.d.ts:31]: Would `AtLeastOne` and `AtLeastN` be better names? Or maybe `MoreThanN`?

Comment (sindresorhus) [reply index.d.ts:31]: > Would AtLeastOne and AtLeastN be better names?

Yes

> Or maybe MoreThanN?

This has a different meaning than `NOrMore`. (`>` vs `>=`)

Comment (sindresorhus) [reply index.d.ts:31]: How about `NonEmpty`?

It could be `NonEmpty` and `MinimumN`.

File: readme.md (hunk 2 above)
Comment (darcyparker) [reply readme.md]: Addressed in 446e3c5

File: index.d.ts (hunk 1 above)
Comment (darcyparker) [reply index.d.ts]: I like `NonEmpty` and `MinimumN`. See 446e3c5.

**Comment by darcyparker** [comment]
FYI: I updated this pull request based on feedback  (and after being MIA for a bit).  I still need to add use cases and a usage example. I will do this after https://github.com/sindresorhus/type-fest/pull/294 is finalized/merged.  Because this draft should be updated to use the updated `FixedLengthArray` (or a new flavour of it is merged to `main`).

File: readme.md [hunk 4]
```diff
@@ -146,6 +146,8 @@ Click the type names for complete docs.
 - [`FixedLengthArray`](source/fixed-length-array.d.ts) - Create a type that represents an array of the given type and length.
 - [`MultidimensionalArray`](source/multidimensional-array.d.ts) - Create a type that represents a multidimensional array of the given type and dimensions.
 - [`MultidimensionalReadonlyArray`](source/multidimensional-readonly-array.d.ts) - Create a type that represents a multidimensional readonly array of the given type and dimensions.
+- [`NonEmpty`](source/minimum-n.d.ts) - Create a type of one or more elements.
+- [`MinimumN`](source/minimum-n.d.ts) - Create a type of N or more elements.
```
Comment (jonahsnider) [inline readme.md:150]: This sounds like it's describing a union type, you should explicitly mention that this is an array type.

File: source/minimum-n.d.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+/**
+Create a type of one or more elements.
+
+@category Utilities
```
Comment (jonahsnider) [inline source/minimum-n.d.ts:4]:
```suggestion
@category Array
```
Categories were made more specific in #331

**Comment by sindresorhus** [comment]
Seeing as this is blocked by https://github.com/sindresorhus/type-fest/issues/284 and that issue is not going anywhere, what should we do about this pull request?

**Comment by darcyparker** [comment]
A solution is now feasible with `ReadonlyTuple` which was implemented https://github.com/sindresorhus/type-fest/pull/383 It is not blocked by #284. 

 I need to come back to this.  I am sorry I have let this sit. I am fine if you kill this PR until I am able to get back to it.

**Comment by sindresorhus** [comment]
Ah ok. That's good news. I can keep the PR open for a couple of months more.

**Comment by sindresorhus** [comment]
Closing for lack of activity.

Moved this into two issues:
- https://github.com/sindresorhus/type-fest/issues/477
- https://github.com/sindresorhus/type-fest/issues/476

**Pull request closed without merging**